capstone/src/main/java/com/google/sps/servlets/UserServlet.java

```diff
@@ -1,11 +1,218 @@
+   * @param response A HttpServletResponse object that holds the information that will be sent back to the user.
+   * @return HashMap<String, String> saves a key-value pair of all neccessary user details derived.
+   */
+  public HashMap<String, String> getUserDetails(
+    DataSource pool,
+    UserService userService,
+    String get_user_row_query,
+    HttpServletResponse response
+  )
+    throws SQLException, IOException {
+    String urlToRedirectToAfterUserLogsOut = "";
+    String urlToRedirectToAfterUserLogsIn = "";
+
+    HashMap<String, String> userDetails = new HashMap<>();
+
+    Boolean isUserLoggedIn = userService.isUserLoggedIn();
+    if (isUserLoggedIn) {
+      String email = userService.getCurrentUser().getEmail();
+      if (!email.endsWith(".edu")) {
+        userDetails.put("status_code", "412");
+        String logoutUrl = userService.createLogoutURL(
+          urlToRedirectToAfterUserLogsOut
+        );
+        userDetails.put("logoutUrl", logoutUrl);
+        return userDetails;
+      }
+      try (Connection conn = pool.getConnection()) {
+        try (
+          PreparedStatement queryStatement = conn.prepareStatement(
+            get_user_row_query
+          )
+        ) {
+          queryStatement.setString(1, email);
+          ResultSet result = queryStatement.executeQuery();
+          if (result.next() == false) {
+            response.sendRedirect("/userDetails.html");
+            userDetails.put("status_code", "307");
+            return userDetails;
+          } else {
+            do {
+              String user_status = Boolean.toString(isUserLoggedIn);
+              userDetails.put("user_status", user_status);
+              String id = Integer.toString(result.getInt("id"));
+              userDetails.put("id", id);
+              userDetails.put("email", email);
+              String full_name = result.getString("full_name");
+              userDetails.put("full_name", full_name);
+              String user_name = result.getString("user_name");
+              userDetails.put("user_name", user_name);
+              String university = result.getString("name");
+              userDetails.put("university", university);
+              String logoutUrl = userService.createLogoutURL(
+                urlToRedirectToAfterUserLogsOut
+              );
+              userDetails.put("logoutUrl", logoutUrl);
+              userDetails.put("status_code", "200");
+            } while (result.next());
+            return userDetails;
+          }
+        }
+      }
+    } else {
+      String user_status = Boolean.toString(isUserLoggedIn);
+      userDetails.put("user_status", user_status);
+      String loginUrl = userService.createLoginURL(
+        urlToRedirectToAfterUserLogsIn
+      );
+      userDetails.put("loginUrl", loginUrl);
+      userDetails.put("status_code", "401");
+      return userDetails;
+    }
+  }
```

Comment:
This whole block is kind of hard to follow mainly due to that there are too many nested blocks (if -> if -> try -> try -> else -> do/while).

You can try to refactor this by putting each separate logic into helper functions and return early when an if condition is not satisfied. Example:
if (!condition) {
 // work
 return
}
// work

will look much nicer than
if (condition) {
  work
} else {
  work
}

Reply:
Done, you can review it.